src/mapreduce.jl

```diff
@@ -159,27 +159,46 @@ end
 (-)(A::DArray, x::Number) = A .- x
 (-)(x::Number, A::DArray) = x .- A
 
-map_localparts(f::Callable, d::DArray) = DArray(i->f(localpart(d)), d)
-map_localparts(f::Callable, d1::DArray, d2::DArray) = DArray(d1) do I
-    f(localpart(d1), localpart(d2))
+function map_localparts(f::Callable, d::DArray)
+    r = Base.asyncmap!(similar(procs(d), Future), procs(d)) do p
+        remotecall_wait(p) do
+            f(localpart(d))
+        end
+    end
+    return DArray(r)
```

Comment:
I don't know if this is a good idea - just throwing it out for discussion.

We could check if `init` is a 1-arg function or a 0-arg function. If 0-arg we assume that it is a closure/function that has all the information required to create localparts (including any number of other DArrays). If 1-arg it is the current form which takes in localindexes as the only argument.

We support the following types of 0-arg-init constructors.

`DArray(0_arg_init_func, pids)` - executes `0_arg_init_func` at all pids and returns a DArray after computing its indexes from the refs. `pids` must be of the same shape and size as the distribution.

`DArray(0_arg_init_func, d::Darray)` - same as above, pids are of the same shape and size as d.

`DArray(0_arg_init_func) = DArray(0_arg_init_func, workers())` will only work if `0_arg_init_func` returns a scalar or vector as localparts. Will distribute across all workers.